### DISCUSSION TITLE
ESLint rewrite request: Arbitrary warning levels.

### DISCUSSION THREAD

**Discussion by dwjohnston** [eslint/eslint#18517] [Ideas]
Currently ESLint only has two warning levels - warn and error. 

Personally I find the making use of the distinction very useful - where red squiggles mean 'something you need to fix now' and yellow squiggles mean 'just fix this before you create a pull request' (eg. no unused vars). 

By default ESLint will exit non-0 if it detects any error level warnings, but we can can configure an non-0 exit for warning level warnings with the `--max-warnings` option. 

In practise `--max-warnings 0` means that neither warnings nor errors are acceptable to pipeline checks, the practical difference being the colour of the squiggle in your IDE. 

**The suggestion** - allow configuring eslint rules at arbitrary levels, and then allow configuration of how eslint responds handles errors at difference levels. 

**Where this might be useful - progressively adopting rules** 

A common scenario I see is a team identifies a code smell in their codebase that could be fixed with an eslint rule. They can add the rule, but this rule is triggered numerous times throughout the codebase, and it's not viable to address all instances of the problem in a big bang PR. 

You can adopt a progressive strategy by doing --max-warnings 100, and progressively reducing the number, but if you have multiple rules at this warning level, they they all fit in the same bucket, it is harder 'how well are we doing at reducing warnings of X type' for example. 

As a second feature request: a first class progressive rule adoption tool.

**Comment by nzakas** [comment]
We've had the request for arbitrary rule severities before, but no one has been able to articulate why that would be valuable. We have "error", which changes the exit code, and "warn" which does not change the exit code. How would you expand that to arbitrary severities? What would be the differentiator?

Warnings were created specifically for the purpose of letting people adopt rules incrementally. You'd see the warning but it wouldn't affect the exit code. I myself did this at my last job, where I let people know that there were some rules we'd start enforcing in a month but for now they are warnings.

So, I'm not sure what "a first class progressive rule adoption tool" actually means. If you have something in mind, please describe it.

**Reply by dwjohnston** [reply]
Let's say you configure your rules a similar fashion to existing eslint, this time instead of just `"off" | "warn" | "error" | 0 | 1 | 2` as valid values, it's any string or number. 

So we have an example rules config like: 

```
"rules": {
   "no-used-vars" : "warn", // Regular existing style 
   "complexity" ["complexity", 2],  // A warning against a 'complexity' tag
   "sort-keys": "sort",   // A warning against a 'sort' tag
   "sort-imports": "sort" // A warning against a 'sort' tag
}
```


Elsewhere in our ESLint config we configure how the arbitrary warnings behave: 

```
"levels": {
    "complexity": {
          "maxAllowed": 0,

          // The maxAllowed property is really what determines the exit status, but presumably we want 
          // Some way to configure whether your IDE shows a red or a yellow squiggle  
          "ideHint": "error",  
    }, 
    "sort": {
          "maxAllowed": 10, 
          "ideHint": "warn"
     }
}
```

If you don't provide a configuration for a given level, I guess default it to how warning behaves. 


Re: Progressive adoption: 

Maybe something like this - you configure the above like: 
```
"levels": {
    "complexity": {
          "maxAllowed": "progressive",
          "ideHint": "error",  
    }, 
    "sort": {
          "maxAllowed": "progressive", 
          "ideHint": "warn"
     }
}
```

Now when you run eslint, it creates a snapshot like: 

```
{
    "complexity": {
         numFound: 50, 
         // Maybe also it tracks the individual locations of the instances and so doesn't allow removing one from one place and adding in another, but maybe that's too hard. 
    }, 
    "sort": {
         numFound: 150, 
    }, 
}

```

Later, if the number of warnings exceeds what's in the snapshot it'll fail, or if it's lower than what's in the snapshot it updates the snapshot.

**Reply by nzakas** [reply]
Hmmm, interesting idea. There was a similar thread here:
https://github.com/eslint/eslint/issues/14679

In general, the team wasn't convinced of this direction. Not sure what the appetite would be now, but you can read through the comments to better understand everyone's position.

Your progressive rollout idea looks a lot like this suppression RFC that's currently in review:
https://github.com/eslint/rfcs/pull/119
